pandas/tests/indexes/datetimes/test_tools.py

```diff
@@ -469,6 +472,42 @@ def test_to_datetime_tz(self, cache):
         with pytest.raises(ValueError, match=msg):
             pd.to_datetime(arr, cache=cache)
 
+    @pytest.mark.parametrize('cache', [True, False])
+    def test_to_datetime_different_offsets(self, cache):
+        # inspired by asv timeseries.ToDatetimeNONISO8601 benchmark
+        # see GH-26097 for more
+        ts_string_1 = 'March 1, 2018 12:00:00+0400'
+        ts_string_2 = 'March 1, 2018 12:00:00+0500'
+        arr = [ts_string_1] * 5 + [ts_string_2] * 5
+        expected = pd.Index([parse(x) for x in arr])
+        result = pd.to_datetime(arr, cache=cache)
+        tm.assert_index_equal(result, expected)
+
+    @pytest.mark.parametrize('errors', ('ignore', 'coerce', 'raise'))
+    @pytest.mark.parametrize('suffix', ([], ['foo']))
+    @pytest.mark.parametrize('convertor', (lambda x: x, str))
+    @given(date1=datetimes(timezones=timezones()),
+           date2=datetimes(timezones=timezones()))
+    def test_to_datetime_cache_invariance(self, date1, date2, suffix,
+                                          errors, convertor):
+        # prepare a list of dates to parse with some duplicates
+        # and possible invalid string
+        arg = [convertor(date1), convertor(date2)] * 5 + suffix
+
+        def _get_answer(cache):
+            try:
+                return pd.to_datetime(arg, cache=cache, errors=errors)
+            except ValueError as err:
+                return err.args
+
+        cache_on = _get_answer(cache=True)
+        cache_off = _get_answer(cache=False)
+        assert type(cache_on) == type(cache_off)
+        if isinstance(cache_on, pd.Index):
```

Comment:
when would this *not* be an Index?

Reply:
When there was a `ValueError` raised within `pd.to_datetime()` call, in this case `cache_on` and `cache_off` must be exception arguments (and must be the same, as the test checks for behaviour being identical with both caching on and off).

Reply:
@jreback latest failure on Travis is ___perfect___ example of why I have added this test:

```
self = <pandas.tests.indexes.datetimes.test_tools.TestToDatetime object at 0x7fe657d14320>
date1 = datetime.datetime(2000, 1, 1, 0, 0, tzinfo=<UTC>)
date2 = datetime.datetime(2000, 1, 1, 0, 0, tzinfo=<StaticTzInfo 'Etc/GMT'>)
suffix = [], errors = 'raise'
convertor = <function TestToDatetime.<lambda> at 0x7fe678d17158>
    @pytest.mark.parametrize('errors', ('ignore', 'coerce', 'raise'))
    @pytest.mark.parametrize('suffix', ([], ['foo']))
    @pytest.mark.parametrize('convertor', (lambda x: x, str))
    @given(date1=datetimes(timezones=timezones()),
           date2=datetimes(timezones=timezones()))
    def test_to_datetime_cache_invariance(self, date1, date2, suffix,
                                          errors, convertor):
        # prepare a list of dates to parse with some duplicates
        # and possible invalid string
        arg = [convertor(date1), convertor(date2)] * 5 + suffix
    
        def _get_answer(cache):
            try:
                return pd.to_datetime(arg, cache=cache, errors=errors)
            except ValueError as err:
                return err.args
    
        cache_on = _get_answer(cache=True)
        cache_off = _get_answer(cache=False)
>       assert type(cache_on) == type(cache_off)
E       AssertionError: assert <class 'pandas.core.indexes.datetimes.DatetimeIndex'> == <class 'tuple'>
E        +  where <class 'pandas.core.indexes.datetimes.DatetimeIndex'> = type(DatetimeIndex(['2000-01-01 00:00:00+00:00', '2000-01-01 00:00:00+00:00',\n               '2000-01-01 00:00:00+00:00', '...       '2000-01-01 00:00:00+00:00', '2000-01-01 00:00:00+00:00'],\n              dtype='datetime64[ns, UTC]', freq=None))
E        +  and   <class 'tuple'> = type(('Tz-aware datetime.datetime cannot be converted to datetime64 unless utc=True',))
pandas/tests/indexes/datetimes/test_tools.py:505: AssertionError
---------------------------------- Hypothesis ----------------------------------
Falsifying example: test_to_datetime_cache_invariance(self=<pandas.tests.indexes.datetimes.test_tools.TestToDatetime at 0x7fe657d14320>, date1=datetime.datetime(2000, 1, 1, 0, 0, tzinfo=<UTC>), date2=datetime.datetime(2000, 1, 1, 0, 0, tzinfo=<StaticTzInfo 'Etc/GMT'>), suffix=[], errors='raise', convertor=lambda x: x)
You can reproduce this example by temporarily adding @reproduce_failure('4.17.2', b'AXicY2DAARgBABsAAg==') as a decorator on your test case
```

Reply:
Given this found failure, what do you think is a correct behaviour (seeing that both `UTC` and `Etc/GMT` seem to be describing same timezone) - should `to_datetime()` be raising an exception or not?

Reply:
Easier reproducer:
```python
import pandas as pd
import datetime
import pytz

t1 = pytz.utc
t2 = pytz.timezone('Etc/GMT')

d1 = datetime.datetime(2000, 1, 1, tzinfo=t1)
d2 = datetime.datetime(2000, 1, 1, tzinfo=t2)

arg = [d1, d2] * 5

print(pd.to_datetime(arg, cache=True, errors='raise'))
print(pd.to_datetime(arg, cache=False, errors='raise'))
```

Reply:
great you can add that as another example

Reply:
But still, what do you feel is correct behaviour for this particular data input - should it raise (like in `cache=False`) or should it return a `DatetimeIndex()` (like in `cache=True`)?

Reply:
After looking more carefully I found slightly less artificial example:
```python
import pandas as pd
import datetime
import pytz

t1 = pytz.timezone('US/Alaska')
t2 = pytz.timezone('Etc/GMT')

d0 = datetime.datetime(2000, 1, 1)
d1 = d0.astimezone(t1)
d2 = d0.astimezone(t2)

arg = [d1, d2]

print(pd.to_datetime(arg, cache=True, errors='raise'))
print(pd.to_datetime(arg, cache=False, errors='raise'))
```
<details>
<summary>script output</summary>

```
>python pd2.py
DatetimeIndex(['1999-12-31 12:00:00-09:00', '1999-12-31 12:00:00-09:00'], dtype='datetime64[ns, US/Alaska]', freq=None)
Traceback (most recent call last):
  File "C:\Vass\Work\cuDF\pandas-dev\pandas\core\arrays\datetimes.py", line 1860, in objects_to_datetime64ns
    values, tz_parsed = conversion.datetime_to_datetime64(data)
  File "pandas\_libs\tslibs\conversion.pyx", line 185, in pandas._libs.tslibs.conversion.datetime_to_datetime64
    raise ValueError('Array must be all same time zone')
ValueError: Array must be all same time zone

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "pd2.py", line 15, in <module>
    print(pd.to_datetime(arg, cache=False, errors='raise'))
  File "C:\Vass\Work\cuDF\pandas-dev\pandas\core\tools\datetimes.py", line 610, in to_datetime
    result = convert_listlike(arg, box, format)
  File "C:\Vass\Work\cuDF\pandas-dev\pandas\core\tools\datetimes.py", line 303, in _convert_listlike_datetimes
    allow_object=True)
  File "C:\Vass\Work\cuDF\pandas-dev\pandas\core\arrays\datetimes.py", line 1865, in objects_to_datetime64ns
    raise e
  File "C:\Vass\Work\cuDF\pandas-dev\pandas\core\arrays\datetimes.py", line 1856, in objects_to_datetime64ns
    require_iso8601=require_iso8601
  File "pandas\_libs\tslib.pyx", line 460, in pandas._libs.tslib.array_to_datetime
    cpdef array_to_datetime(ndarray[object] values, str errors='raise',
  File "pandas\_libs\tslib.pyx", line 537, in pandas._libs.tslib.array_to_datetime
    raise ValueError('Tz-aware datetime.datetime '
ValueError: Tz-aware datetime.datetime cannot be converted to datetime64 unless utc=True

```
</details>

After looking into this example I'm inclined to declare `cache=False` behaviour as correct (and thus `cache=True` has a bug). Ideas?

Reply:
you can do this in "script output" summary above.
or do we need to pull information from this summary and insert here?

Reply:
@jreback problem in a nutshell - if you have `datetime` objects with dates that are equal but lay in different timezones then `cache=True` and `cache=False` behaviour differs (because for `cache=True` Pandas first compute unique subset, and that is done by comparing objects, and dates are considered equal, so they're translated to same object with timezone equal to the first one).

Also the output is shown when you click on "show output" in my comment above (I've copied it down below in case you're reading from something like e-mail):

```
>python pd2.py
DatetimeIndex(['1999-12-31 12:00:00-09:00', '1999-12-31 12:00:00-09:00'], dtype='datetime64[ns, US/Alaska]', freq=None)
Traceback (most recent call last):
  File "C:\Vass\Work\cuDF\pandas-dev\pandas\core\arrays\datetimes.py", line 1860, in objects_to_datetime64ns
    values, tz_parsed = conversion.datetime_to_datetime64(data)
  File "pandas\_libs\tslibs\conversion.pyx", line 185, in pandas._libs.tslibs.conversion.datetime_to_datetime64
    raise ValueError('Array must be all same time zone')
ValueError: Array must be all same time zone

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "pd2.py", line 15, in <module>
    print(pd.to_datetime(arg, cache=False, errors='raise'))
  File "C:\Vass\Work\cuDF\pandas-dev\pandas\core\tools\datetimes.py", line 610, in to_datetime
    result = convert_listlike(arg, box, format)
  File "C:\Vass\Work\cuDF\pandas-dev\pandas\core\tools\datetimes.py", line 303, in _convert_listlike_datetimes
    allow_object=True)
  File "C:\Vass\Work\cuDF\pandas-dev\pandas\core\arrays\datetimes.py", line 1865, in objects_to_datetime64ns
    raise e
  File "C:\Vass\Work\cuDF\pandas-dev\pandas\core\arrays\datetimes.py", line 1856, in objects_to_datetime64ns
    require_iso8601=require_iso8601
  File "pandas\_libs\tslib.pyx", line 460, in pandas._libs.tslib.array_to_datetime
    cpdef array_to_datetime(ndarray[object] values, str errors='raise',
  File "pandas\_libs\tslib.pyx", line 537, in pandas._libs.tslib.array_to_datetime
    raise ValueError('Tz-aware datetime.datetime '
ValueError: Tz-aware datetime.datetime cannot be converted to datetime64 unless utc=True
```

Reply:
this is not a reasonable check here, use assert_almost_equal(cache_on, cache_off)

the point is they should be identical.

Reply:
Done